components/merkledb/src/views/metadata.rs

```diff
@@ -346,13 +345,13 @@ impl<T: RawAccess> IndexesPool<T> {
         (metadata, is_phantom)
     }
 
-    pub(crate) fn suffixes<K: BinaryKey + ?Sized>(&self, prefix: &str) -> Vec<K::Owned> {
-        let prefix = concat_keys!(prefix, INDEX_NAME_SEPARATOR);
+    pub(crate) fn suffixes<K: BinaryKey + ?Sized>(&self, prefix: &IndexAddress) -> Vec<K::Owned> {
+        let prefix = prefix.qualified_prefix();
         let mut keys = Vec::new();
+        let mut bytes_iter = self.0.iter_bytes(&prefix);
 
-        for (key, _) in self.0.iter::<_, Vec<u8>, Vec<u8>>(&prefix) {
-            let split = key.split_at(prefix.len());
-            let suffix = K::read(&split.1);
+        while let Some(key) = bytes_iter.next() {
```

Comment:
`bytes_iter` doesn't stop when the generated keys no longer start with `prefix`, so this must be checked manually:

```rust
while let Some((key, _)) = bytes_iter.next() {
    if !key.starts_with(&prefix) {
        break;
    }
    // ...
}
```

Note that this leads to incorrect values returned from `buffered_keys`, possible panics (if one of the following keys is shorter than `prefix`) and undefined behavior (if the key type is `str` and a successive key is not valid UTF-8). Unfortunately, the tests as written don't capture this; IMO, they should capture all these situations.